### PR TITLE
Various Reason syntax fixes

### PR DESCRIPTION
**Pull request by glennsl**
There's also a lot of punctuation that's either uncategorized or miscategorized, and the categorization in general seems extremely hacky. Made to fit some syntax theme(s) perhaps? If so, that seems really backwards since themes can and should have rules targeting specific grammars. But as it is, any change towards sanity will probably break something, so I've made the commits pretty fine-grained in order to isolate and justify the changes better.

### CODE REVIEW

**Comment by freebroccolo** [comment]
First, thanks for working on this.

Regarding the "miscategorization" of tokens, yes, it is intentional. I have commented on this in a few places, including [here](https://github.com/reasonml-editor/vscode-reasonml/issues/61#issuecomment-338504085).

The short version is that using the de-facto TextMate scopes leads to sub-optimal highlighting. These scopes were never standardized and were mostly chosen in an ad-hoc fashion over time to suit languages like C and Ruby.

As a result, the scopes simply do not map well onto the syntactic categories that the OCaml grammar uses. There just aren't distinctions in the de-facto TextMate scopes for all the syntactic categories like pattern variables, value constructors, type constructors, record paths, module paths, etc.

What I have done is to try and follow the de-facto scopes up to a point, after which I have relied on a process of hand tuning the scopes for several major themes across atom, code, etc.

I find that this approach leads to better highlighting (providing more useful visual distinction) but the downside is that the highlighting may appear less consistent in comparison to other languages when using the same theme.

I think this is an acceptable trade off but I realize others may disagree.

In any case, because this has come up a number of times recently, I am considering switching to a simpler and (in comparison to other language highlighting) perhaps more consistent version of the highlighter in this repository and just providing a more advanced one separately.

Personally, I think it would be a bit of a shame to do that because a lot of people will just install the extension and not even realize there is the possibility of more fine-grained and more accurate highlighting.

**Comment by glennsl** [comment]
Thanks for the info. I understand that the conventional TextMate scopes are insufficient, but it seems possible to at least broadly tag most tokens in a way that is decently well supported by most syntax themes. That still leaves it sub-optimal, as you rightly point out, but optimization should then be done in each theme, not in the grammar. Otherwise you optimize for a handful of themes to the detriment of all others, and the ecosystem as a whole, since now themes will need hacks to counter your hacks, which will eventually come around to bite you. And so it goes.

Here's a few examples of issues I've encountered that make proper highlighting difficult:
* Keywords like `let`, `module` and `type` are tagged as language-defined types instead of as keywords (fixed in this PR).
* Pattern match vairables are tagged as both a hyperlink and a reserved variable, instead of as an ordinary variable (fixed in this PR)
* Variant constructors, booleans and such are categorized as numeric constants, thereby making it impossible to distinguish actual number literals from the other tokens (unless perhaps you invert the problem and distinguish the other tokens from number literals. Hacks to counter hacks, you see?).
* Types are tagged as `string.regexp` in addition to `support.type`, which makes it highlight as a string (I'm not entirely sure why, since `support.type` should be more specific than just `string`, but for some reason the latter wins)
* Semicolon is (often?) tagged as `variable.other.class.js variable.interpolation keyword.operator keyword.control message.error` instead of as punctuation. This seems to be a deeper issue than just miscategorization since it's not being matched alone, but the effect is still that it's impossible to highlight punctuation properly (also, most other punctuation is simply untagged so that'd still have to be fixed of course).

So to be clear, removing the hacks doesn't have to mean sub-optimal highlighting. It just means you have to optimize where you should, in each theme. Which might be cumbersome, but is better in the long run and for the wider ecosystem, outside of just the 5 or so themes you've tested and hand-tuned it for.

I any case, if there's part of this PR you'd accept, but others you'd like to hold off on or change, please let me know. A little improvement is better than nothing :)

**Comment by freebroccolo** [comment]
Thanks for the detailed response.

I had written up my own longer response to this, including clarifications and questions, but due to the fact that I'm going to be short on time it might be best to table an in-depth discussion for now.

What I will say is that most of your changes appear to remove intentional visual distinctions (the "miscategorizations") in favor of sticking close to the intended semantic categories for certain scopes. 

I do not have screenshots to compare but I suspect I would find these changes to result in less informative highlighting even if (according to the scope names) it is "more proper" in the sense you are suggesting.

I think this is fundamental a philosophical disagreement: I value informative colorization more than consistent colorization (in comparison to how other languages are colorized). From my perspective, the semantic intent of existing scopes is mostly worthless and their only purpose is to provide colorization in an approximately consistent fashion across major themes.

I realize that means this extension does not integrate into the ecosystem as nicely as one would want. However, I don't think that the suggestion (fixing the themes) is very feasible. Even fixing 3 or 4 major themes across the major editors will require a lot of work. It will also make changes over time extremely difficult to coordinate. Then there's the problem of trying to get changes to the default themes (for atom and code) accepted upstream.

I just don't see it happening, but I admit I haven't tried. But if someone else really wants to take that up and push for it, I'd be willing to support their efforts.

As far as the future is concerned, it sounds like at least Atom and Github are going to be moving toward [tree-sitter](https://github.com/tree-sitter). I am hoping this will provide an opportunity to reset the current situation with scopes and themes somehow. I would much rather put effort in that direction rather than continue trying to optimize (or even fix) the current highlighting in the sense you are concerned with.

tl;dr: I'd prefer to accept the changes for pattern guards but leave the others on the table for now, with the intention to come back to it once.

**Comment by glennsl** [comment]
Alright, thanks for the thorough reponse. Unfortunately it still doesn't seem like you understand what I'm trying to say with regards to how this affects themes other than those you personally prefer. 

I think I'll just maintain a fork instead then. Feel free to do whatever you wish with this PR.

**Comment by freebroccolo** [comment]
> Unfortunately it still doesn't seem like you understand what I'm trying to say with regards to how this affects themes other than those you personally prefer.

@glennsl I don't really prefer any of the themes. I have tuned for specifically for the default themes for Atom and VS Code and then several of the most popular themes in the package repositories.

I think I understand what you are asking for: that we stick to de-facto scopes as much as possible and  push language-specific tuning out into the themes. I agree that in principle this would be the better solution. In practice, I don't think it's going to be feasible to maintain changes to several themes that are spread across many different repositories outside of our control. Not only that, but we'd need to be able to convince the editor devs to include OCaml/Reason specific optimizations in the default themes.

But lets set all of that aside because I think the concrete contributions of this PR have gotten lost in a philosophical discussion.

Are you willing to post some screenshots that show the effect of your changes so that there is something more specific to talk about? I just want to make sure we are on the same page here.

**Comment by glennsl** [comment]
Sorry, it'd probably be more accurate to say "your preferred variations on a select few themes". My point is just that there's a very small set of themes you'd be able to maintain this way, And only _you_ will be able to maintain them since they're tuned to your preference. It makes no sense to anyone else.

I really cannot see how decentralization would be less feasible than the current approach,. which doesn't scale at all. While it's true that it might be difficult to get optimizations upstreamed, maintaining and recommending a fork doesn't seem like an unreasonable option. These themes usually don't change a whole lot. Even for the small set of themes you have optimized for today, it seems more feasible to maintain it in such a way that other people can understand and be able to share the burden.

I did check the `Dark+` theme to see what the changes did, by the way, and while of course it did change, it looked fine to me. But whether it suits your preference is anyone's guess but yours.

While I'm sure it's possible to find a solution to this concrete PR that would satisfy us both, I have very little interest in trying to work with the current approach. The PR is just what I managed to get done before I gave up trying. I'd really rather maintain a fork of the grammar instead. It's not ideal, but I'm fine with that. I don't mean to force you into making a decision either. Please take your time to think about it, and if you do change your mind I'm sure we can consolidate then.

And again, feel free change and merge what you want in this PR. I really don't mind.

**Comment by freebroccolo** [comment]
Thanks for the response. I think I will close this for now but I appreciate the effort you put into it. My plan, once I get some time to rewrite the Reason syntax highlighting, is to just simplify the scopes. If I have time I may try to maintain an advanced theme.
